Make title element more human-readable

mint-thompson opened the pull request:
The title attributes on structure definitions, extensions, and logical models make use of
a new "title" function on the Identifier class. This function uses the Identifier's name and
inserts spaces based on word boundary conditions. Because a title may contain letters
and numbers, it is necessary to use somewhat unusual-looking statements when
checking for word boundaries.

This pull request addresses https://standardhealthrecord.atlassian.net/projects/CIMPL/issues/CIMPL-79

For testing, I used the mCODE spec. The following elements have names that were useful in testing:
- Basophils100WBCNFrPtBldQnAutoCntLabObs
- PromyelocytesNCncPtBldQnLabObs
- ComprehensiveMetabolic2000SerumOrPlasmaPanel
- GFR173sqMPredNonBlackArVRatPtSerPlasBldQnMDRDLabObs

mint-thompson committed:
Make title element more human-readable

The title attributes on structure definitions, extensions, and logical models make use of
a new "title" function on the Identifier class. This function uses the Identifier's name and
inserts spaces based on word boundary conditions. Because a title may contain letters
and numbers, it is necessary to use somewhat unusual-looking statements when
checking for word boundaries.

ngfreiter commented:
One thing to take a look at is it looks like this change caused some (sadly a lot) of tests to fail because they all used the old naming style. It didn't look like it was anything more complicated than that, but we will need to update those tests.

mint-thompson committed:
Update test fixtures and deserialization for FHIR export

kjmahalingam commented on packages/shr-models/lib/models.js:
```diff
@@ -754,6 +754,40 @@ class Identifier {
   get namespace() { return this._namespace; }
   get name() { return this._name; }
   get fqn() { return (this.isPrimitive || this.isSpecialKeyWord) ? this._name : `${this._namespace}.${this._name}`; }
+  get title() {
```
I don't have time to review this in too much detail right now, but I just want to say that I appreciate that you wrote this out in a logical way distinct from the regular expression I previously used elsewhere. This, although larger, will likely be more easily maintainable than that regex.

Assuming that someone can test this and it works well, good job! And good job in progress, otherwise. :)

cmoesel commented:
Congratulations on having PR #1!

I haven't tested this yet, but the code looks good!  My only suggestion would be to use it in more places.  For example, I think we'd want it for:
* ValueSet: https://github.com/standardhealth/shr-tools/blob/master/packages/shr-fhir-export/lib/valueSets.js#L39
* CodeSystem: https://github.com/standardhealth/shr-tools/blob/master/packages/shr-fhir-export/lib/codeSystems.js#L38

We should also probably update the Data Dictionary exporter to use this instead of the regular expressions that Dylan was using.

mint-thompson committed:
Additional use of Identifier title in FHIR and Data Dictionary exports

In FHIR export, use Identifier title in code systems and value sets.
In Data Dictionary export, use Identifier title for profile and data element names.
Note that the Data Dictionary does not use the Identifier title for Value Sets.
The regular expressions previously used to create the title are removed.

cmoesel commented:
Looking good!  I ran it and the profiles/extensions/valuesets/codesystems/datadict came out great.  There's one last thing I found though...  When you render the IG, the Profiles and Extensions pages are still listing using the `name` instead of the `title`.  For example:

![image](https://user-images.githubusercontent.com/2278253/66424480-b547c500-e9db-11e9-9157-b9b5c7f1806e.png)

I think we should list the title here.  It looks like it's correct on ValueSets and CodeSystems pages, but probably needs to be updated on Profiles, Extensions, and Logical Models pages.

Those pages get built in the shr-fhir-exporter ig.js file.

cmoesel commented:
Note that when you get the titles in the ig.js you'll need to use the `MVH.sdTitle` and `MVH.vsTitle` functions since the title property changed its name from DSTU2 to STU3/R4.

cmoesel committed:
Use title when generating IG listing pages

cmoesel reviewed (approved):
Great work!  I especially like that last commit! ;-)
